pallets/pallet-bank/src/lib.rs

```diff
@@ -267,6 +433,59 @@ impl<T: Config> BasicAccounting<T::AccountId, T::Balance> for Pallet<T> {
     }
 }
 
+impl<T: Config> Stakable<T::AccountId, T::Balance> for Pallet<T> {
+    fn stake_funds(user: &T::AccountId, amount: T::Balance) -> DispatchResult {
+        // Stake funds from free to reserved
+        T::RoleManager::ensure_role(&user, Role::Customer)?;
+        Accounts::<T>::mutate(&user, |balance| -> DispatchResult {
+            if balance.free >= amount {
+                balance.free -= amount;
+                balance.reserved += amount;
+                Ok(())
+            } else {
+                Err(Error::<T>::InsufficientBalance.into())
+            }
+        })?;
+        Self::deposit_event(Event::<T>::Stake {
+            user: user.clone(),
+            amount,
+        });
+
+        Ok(())
+    }
+
+    fn redeem_funds(user: &T::AccountId, amount: T::Balance) -> DispatchResult {
+        // Redeem funds from reserved to free after a certain time
+        // get unlock BlockNumber
+        let unlock = T::BlockNumberProvider::current_block_number() + T::RedeemPeriod::get();
+
+        // Add new locked funds to user's Account Data
+        Accounts::<T>::mutate(&user, |balance| -> DispatchResult {
```

Comment:
better call this "account" instead of "balance". Balance usually represents a number, account can represent more complex type